EditorManager

Matthieu-Riou opened the pull request:
Add a manager that allow multiple editors, clipboard and save/load

Matthieu-Riou committed:
A first manager

Matthieu-Riou committed:
Add methods interacting with the current Editor

Matthieu-Riou committed:
Corrections

Matthieu-Riou committed:
Add some test units

Cowa commented on src/main/scala/io/atal/butterfly/EditorManager.scala:
```diff
@@ -0,0 +1,79 @@
+    */
+  def openEditor: Unit = {
+    _editors = new Editor() :: _editors
+    _currentEditor = Some(_editors.head)
```
Les attributs préfixés `_` ne doivent pas être utilisés (même si ça appartient à la classe), sauf dans leur getter/setter bien sûr, ça rend les choses plus jolies et propre :)

Matthieu-Riou replied:
Sauf bien sûr quand on a pas de getter :stuck_out_tongue_closed_eyes: 
Ok, je vais changer ça

Matthieu-Riou committed:
A usefull method for cursors

Matthieu-Riou committed:
Corrections

Matthieu-Riou committed:
The method already exists, my bad.
Do you want to keep my fonctional version ?

Matthieu-Riou committed:
Add Actions

Matthieu-Riou commented:
Last commit refears #18

Matthieu-Riou committed:
Better actions

Matthieu-Riou committed:
Better actions, change the erase of selections

Matthieu-Riou committed:
Remove useless (and unsecure) methods

Matthieu-Riou committed:
A selection is now related to a cursor

Matthieu-Riou committed:
EditorManager call Action instead of Editor. Implements Copy

Matthieu-Riou committed:
Changing Actions
An Action should not directly modify the editor. It calls an editor's method witch modify itself

Cowa commented on src/main/scala/io/atal/butterfly/Action/Action.scala:
```diff
@@ -0,0 +1,7 @@
+/** Define an Action
+  */
+trait Action {
+  def execute(editor: Editor, clipboard: Clipboard): Unit
```
C'est quand même bien moche de devoir passer le Clipboard dans cette méthode, mais je sais bien qu'on est obligé par rapport à `Copy` et `Cut`.
A moins qu'on mette une référence du Clipboard de EditorManager dans Editor également. Mais bon, ça serait de la bidouille sans vraiment de sens.

Donc autant laisser comme ça :+1:

Matthieu-Riou replied:
Yep, et ça a plus de sens : une commande s'exécute (peut s'exécuter) sur un editor et un clipboard donnés.

Matthieu-Riou commented:
Mise en place propre du pattern Commande : 
Les commandes sont appelée depuis EditorManager (qui leur fournit l'Editor courant, et le Clipboard).
Une commande ne modifie pas directement l'Editor, elle appelle une méthode et le laisse se modifier lui même.

Méthodes de base d'Editor : 
- write 
- erase
- moveCursor(Left/Right/Up/Down)
- moveSelection
- clearSelection
- select
- getSelectionContent

Commande de base d'Editor : 
- Write
- Erase
- AddCursor
- MoveCursor(Left/Right/Up/Down)
- MoveSelection
- ClearSelection

Les commandes Select et GetSelectionContent ne peuvent pas exister : une commande n'a pas vocation à retourner quelque chose. 

Commande de base avec Clipboard : 
- Copy
- Paste

Commande Composite de Base : 
- Insert(Position) => MoveCursor + Write
- Erase(Position) => MoveCursor + Write
- Cut => Copy + Erase
- SelectAll(String) => (content.find(string)) + MoveCursor AND AddCursor + MoveSelection

Commande Composite sur Composite : 
- EraseAll(String) => SelectAll(string) + Erase

Etc,etc...

Qu'en dis-tu ?

Cowa commented:
Je suis...
.
.
.
.
.

.
.
.
D'accord !
Ca me semble être la meilleure solution, ça répond bien à la demande, tout en restant assez élégant .

Matthieu-Riou commented:
Cool :+1:

Matthieu-Riou committed:
Intern methods should be private

Matthieu-Riou committed:
Add actions to move cursors

Matthieu-Riou committed:
Add Cut and Paste

Matthieu-Riou committed:
Correct comments
